Documentation/Reproducible-MSBuild/Techniques/TargetFrameworkRootPath.md

```diff
@@ -0,0 +1,46 @@
+# TargetFrameworkRootPath
+
+The "GetReferenceAssemblyPaths" target in "Microsoft.Common.CurrentVersion.targets" resolves from Program Files and Windows Registry where it will search for .net framework reference assemblies (e.g. `<Reference>`).  Instead, recommend utilizing the Microsoft.NETFramework.ReferenceAssemblies nuget instead. Version 1.0.0 of this nuget will provide framework references for all desktop framework versions. 
+
+The `TargetFrameworkRootPath` property lets the user override this process and choose a different location to provide the target framework. By overriding this, we can prevent accidentally picking up target framework installations from registry or program files locations.
+
+- Recommendation: Always
+- Impact: Low
+
+## Usage
+
+In Directory.Build.props
+
+```xml
+<!-- Disable the normal search path mechanism -->
+<PropertyGroup>
+  <TargetFrameworkRootPath Condition="'$(BuildingInsideVisualStudio)'!='true'">[UNDEFINED]</TargetFrameworkRootPath>
+</PropertyGroup>
+<Target Name="CheckTargetFrameworkRootPath" BeforeTargets="GetReferenceAssemblyPaths" Condition="'$(BuildingInsideVisualStudio)'!='true'">
+  <Warning
+    Condition="'$(TargetFrameworkRootPath)' == '[UNDEFINED]' and '$(TargetFrameworkIdentifier)' == '.NETFramework'"
+    Text="Error, TargetFrameworkRootPath not initialized. If you're building for net462 or any other version of desktop NETFramework, please reference the 'Microsoft.NETFramework.ReferenceAssemblies' nuget package and run restore on the project to fix up your framework reference paths." />
+</Target>
+```
+
+Alternatively, you may opt to force importing of the Microsoft.NETFramework.ReferenceAssemblies package. However, this should only be performed in repos that are no longer using packages.config
+
+```xml
+<!-- Disable the normal search path mechanism -->
+<PropertyGroup>
+  <TargetFrameworkRootPath Condition="'$(BuildingInsideVisualStudio)'!='true'">[UNDEFINED]</TargetFrameworkRootPath>
+</PropertyGroup>
+<ItemGroup>
+  <PackageReference Include="Microsoft.NETFramework.ReferenceAssemblies" Version="1.0.0" VersionOverride="1.0.0" />
```

Comment:
Any thoughts on how to do this in a CentralPackageVersions-agnostic way? Specifying Version will error if Microsoft.Build.CentralPackageVersions is used, but omitting it and only specifying VersionOverride will result in warnings if the user isn't using CPV.

Any any thoughts on how we ought to do this if we move this technique into a nuget package?

Reply:
It's possible to version-less includes within the projects and then have a Directory.Build.targets file that does a `<PackageReference Update="Microsoft.NETFramework.ReferenceAssemblies" Version="1.0.0" />` so that it sets the given metadata on the items that are already there.

That lets you easily use Update to specifiy versions once.

Another approach is to use a property like this `<PackageReference Update="Microsoft.NETFramework.ReferenceAssemblies" Version="$(ThePackageVersion)"  />` and then define the properties once.

Reply:
Understood, but in all those cases, this requires manual intervention on the part of the user. Is there any code snippet that has a reasonable chance of working in most reasonable repo configurations?

Reply:
I'm not sure. I tend to use Dependabot for bumping the versions and it's been pretty good about detecting them.